### PR TITLE
Accelerate Unsafe CAS Intrinsics on Aarch64

### PR DESCRIPTION
**Pull request by IBMJimmyk**
Adds support for the following recognized methods:
```
CompareAndExchangeObject     //JDK11
CompareAndExchangeReference  //JDK17,21
CompareAndExchangeInt        //JDK11,17,21
CompareAndExchangeLong       //JDK11,17,21
```

The accelerated CAE code was built on top of the existing accelerated CAS support on Arm. Changes are similar to previously added support on Power, X and Z.

Cleans up `isUnsafeCAS` to remove the now unused `comp` parameter. Also cleans up `isUnsafeWithObjectArg`, `isVarHandleAccessMethod` and `isSignaturePolymorphicMethod` to remove unused comp parameters.

Edit:
Needs to be check in with:
https://github.com/eclipse/omr/pull/7505
Due to the changes to the parameters of `isUnsafeCAS` and `isUnsafeWithObjectArg`.

### CODE REVIEW

**Comment by IBMJimmyk** [comment]
@hzongaro When you get a chance, can you look over the common changes here as well?

**Comment by IBMJimmyk** [comment]
@knn-k I haven't had to submit Arm changes before. Are you the one I need to ask to review Arm changes or is there someone else?

**Review by hzongaro — APPROVED** [review]
I think the changes in common code look good.  Thanks!

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -4836,7 +4847,16 @@ genCAS(TR::Node *node, TR::CodeGenerator *cg, TR_ARM64ScratchRegisterManager *sr
       op = casWithoutSync ? (is64bit ? TR::InstOpCode::casx : TR::InstOpCode::casw) : (is64bit ? TR::InstOpCode::casalx : TR::InstOpCode::casalw);
       generateTrg1MemSrc1Instruction(cg, op, node, resultReg, TR::MemoryReference::createWithDisplacement(cg, addrReg, 0), newVReg);
       generateCompareInstruction(cg, node, resultReg, oldVReg, is64bit);
-      generateCSetInstruction(cg, node, resultReg, TR::CC_EQ);
+
+      if (!isExchange)
+         {
+         generateCSetInstruction(cg, node, resultReg, TR::CC_EQ);
+         }
+      else if (isReference && comp->target().is64Bit() && comp->useCompressedPointers())
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:4855]: No need for `is64Bit()` here.  AArch64 always runs in the 64-bit mode.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -4871,7 +4891,16 @@ genCAS(TR::Node *node, TR::CodeGenerator *cg, TR_ARM64ScratchRegisterManager *sr
       else
          generateCompareImmInstruction(cg, node, resultReg, oldValue, is64bit);
       if (!createDoneLabel)
-         generateTrg1ImmInstruction(cg, TR::InstOpCode::movzx, node, resultReg, 0); // failure
+         {
+         if (!isExchange)
+            {
+            generateTrg1ImmInstruction(cg, TR::InstOpCode::movzx, node, resultReg, 0); // failure
+            }
+         else if (isReference && comp->target().is64Bit() && comp->useCompressedPointers())
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:4899]: The same comment as above.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -4882,72 +4911,83 @@ genCAS(TR::Node *node, TR::CodeGenerator *cg, TR_ARM64ScratchRegisterManager *sr
...
 
       if (createDoneLabel)
          {
          generateLabelInstruction(cg, TR::InstOpCode::label, node, doneLabel);
-         generateCSetInstruction(cg, node, resultReg, TR::CC_EQ);
+         if (!isExchange)
+            {
+            generateCSetInstruction(cg, node, resultReg, TR::CC_EQ);
+            }
+         else if (isReference && comp->target().is64Bit() && comp->useCompressedPointers())
```
Comment (knn-k) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:4927]: The same comment as above.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 1 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:4855]: Please add description of the `isExchange` case in the comment above (lines 4840-4845).
I would appreciate if you also rewrite `oldReg` at line 4843 to `oldVReg`.

Comment (IBMJimmyk) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:4855]: I can change that.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 2 above)
Comment (IBMJimmyk) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:4899]: I'll change that.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp (hunk 3 above)
Comment (IBMJimmyk) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp:4927]: I'll change that as well.

**Comment by knn-k** [comment]
Jenkins test sanity alinux64,amac jdk17 depends eclipse/omr#7505

**Comment by knn-k** [comment]
Jenkins test sanity.system amac jdk17 depends eclipse-omr/omr#7505

**Comment by knn-k** [comment]
Jenkins compile xlinux,plinux jdk21 depends eclipse-omr/omr#7505

**Review by knn-k — APPROVED** [review]
LGTM.

**Comment by knn-k** [comment]
I tried to perform a coordinated merge with https://github.com/eclipse-omr/omr/pull/7505, but the Mirror-OMR-To-OpenJ9-OMR job in the procedure failed unexpectedly.

**Comment by pshipton** [comment]
It's fixed now.

**Comment by knn-k** [comment]
Merging now.

**Pull request merged**